crates/optimism/txpool/src/supervisor/metrics.rs

```diff
@@ -0,0 +1,123 @@
+    pub(crate) cross_unsafe: Histogram,
+
+    /// Frequency of cross chain transactions failing validation due to being
+    /// [`LocalSafe`](SafetyLevel::LocalSafe).
+    pub(crate) local_safe: Histogram,
+
+    /// Frequency of cross chain transactions failing validation due to being
+    /// [`Safe`](SafetyLevel::Safe).
+    pub(crate) safe: Histogram,
+}
+
+impl FailedSafetyLevelMetrics {
+    /// Updates metrics for a batch of transactions from given intermediary [`SafetyLevelCounter`].
+    pub(crate) fn update_batch(&self, safety_level_counter: &SafetyLevelCounter) {
+        let SafetyLevelCounter { invalid, r#unsafe, cross_unsafe, local_safe, safe } =
+            safety_level_counter;
+
+        self.invalid.record(invalid.swap(0, Ordering::Relaxed) as f64);
+        self.r#unsafe.record(r#unsafe.swap(0, Ordering::Relaxed) as f64);
+        self.cross_unsafe.record(cross_unsafe.swap(0, Ordering::Relaxed) as f64);
+        self.local_safe.record(local_safe.swap(0, Ordering::Relaxed) as f64);
+        self.safe.record(safe.swap(0, Ordering::Relaxed) as f64);
+    }
+}
+
+/// Helper type for batch updating [`FailedSafetyLevelMetrics`].
+///
+/// Counts number of cross chain transactions, by their [`SafetyLevel`], that
+/// fail validation due to not meeting locally configured minium safety level requirement.
+#[derive(Debug, Default)]
+pub struct SafetyLevelCounter {
+    /// Count of cross chain transactions failing validation due to being
+    /// [`Invalid`](SafetyLevel::Invalid).
+    pub(crate) invalid: AtomicUsize,
+
+    /// Count of cross chain transactions failing validation due to being
+    /// [`Unsafe`](SafetyLevel::Unsafe).
+    pub(crate) r#unsafe: AtomicUsize,
+
+    /// Count of cross chain transactions failing validation due to being
+    /// [`CrossUnsafe`](SafetyLevel::CrossUnsafe).
+    pub(crate) cross_unsafe: AtomicUsize,
+
+    /// Count of cross chain transactions failing validation due to being
+    /// [`LocalSafe`](SafetyLevel::LocalSafe).
+    pub(crate) local_safe: AtomicUsize,
+
+    /// Count of cross chain transactions failing validation due to being
+    /// [`Safe`](SafetyLevel::Safe).
+    pub(crate) safe: AtomicUsize,
+}
+
+impl SafetyLevelCounter {
+    /// Increases counter for given [`SafetyLevel`].
+    pub(crate) fn increase_safety_level_count_once_for(&self, level: SafetyLevel) {
+        if matches!(level, SafetyLevel::Invalid) {
+            self.invalid.fetch_add(1, Ordering::Relaxed);
+        } else if matches!(level, SafetyLevel::Unsafe) {
+            self.r#unsafe.fetch_add(1, Ordering::Relaxed);
+        } else if matches!(level, SafetyLevel::CrossUnsafe) {
+            self.cross_unsafe.fetch_add(1, Ordering::Relaxed);
+        } else if matches!(level, SafetyLevel::LocalSafe) {
+            self.local_safe.fetch_add(1, Ordering::Relaxed);
+        } else if matches!(level, SafetyLevel::Safe) {
+            self.safe.fetch_add(1, Ordering::Relaxed);
+        }
```

Comment:
nit - Even though the metrics aren't exhaustive over all variants of `SafetyLevel`, prefer an exhaustive `match` w/ `SafetyLevel::Finalized` panicking (or ignored explicitly; handling this higher up is probably preferable.) If we ever get a `SafetyLevel::Finalized` here from the minimum safety error, something's gone very wrong or the function is being misused.